Use GitHub Repo Image registry instead of ECR

NathanielRN opened the pull request:
# Description

~In https://github.com/aws-observability/aws-otel-python/pull/26 we removed the GitHub Registry publishing step because Soak Tests shouldn't be pushing Images.~

~In this PR, we add it back as a part of the Integration Test workflow. In the future, workflows like the Soak Tests could pull the image published by the Integration Test.~

Constantly creating images of the Sample App using the latest packages from OpenTelemetry Python upstream helps us validate our AWS components still work with any new changes. Currently, we publish those images ECR.

Publishing to ECR is a problem because it requires an AWS account to publish the image. Further, viewing the images requires going to `public.ecr.aws` which is less obvious. In this PR we remove those pain points by using GHCR (GitHub Container Registry) which only requires a `GITHUB_TOKEN` with the `packages: write` permissions to upload images. Then, these images are accessible directly from the repo homepages for anyone interested in seeing more information about.

anuraaga commented on .github/workflows/integration-testing.yml:
```diff
@@ -50,13 +54,19 @@ jobs:
+          echo "MATRIX_COMMIT_COMBO=${{ matrix.app-platform }}-${{ matrix.instrumentation-type }}-${{ github.sha }}" | tee --append $GITHUB_ENV;
+      - name: Construct Sample App image tag
+        run: |
+          echo "APP_IMAGE=ghcr.io/${{ github.repository }}-sample-app-${{ env.MATRIX_COMMIT_COMBO }}:latest" | tee --append $GITHUB_ENV;
```
A slash needs to follow the repository so it's scoped to this repository.

NathanielRN replied:
Cool thanks for that, changes made.

anuraaga reviewed (commented):
While I think switching from ECR to GHCR could provide improvements in allowing stable pulls without authentication, does this have anything to do with #26?

Also to clarify I've been thinking of trying GHCR but think it might be possible the org settings don't actually allow it. Hopefully they do but it's possible it won't actually work right now.

NathanielRN commented:
@anuraaga 

> does this have anything to do with #26?

The relationship to #26 is that this PR adds _back_ the publishing of Docker Image Snapshots to GHCR (which #26 removed). You can see this by the [packages which were already published](https://github.com/orgs/aws-observability/packages?repo_name=aws-otel-python) on this repo during the 2 days the Soaking Tests were configured to do so.

I jumped on making this PR because I did not want us to lose the code needed to push to GHCR now that Soak Tests have been updated to not publish anywhere (even GHCR).

> Also to clarify I've been thinking of trying GHCR but think it might be possible the org settings don't actually allow it. Hopefully they do but it's possible it won't actually work right now.

Are org restrictions something we can observe at the repository level? Like I mentioned, we [already have packages published to this repo](https://github.com/orgs/aws-observability/packages?repo_name=aws-otel-python) from when they were temporarily published by the Soak Tests. I guess this means we do not have restrictions? Maybe the restrictions did not trigger because I did not have the slash in the image name that you mention in https://github.com/aws-observability/aws-otel-python/pull/28#discussion_r714432265?

anuraaga commented:
It looks like hyphen works too similarly to a slash, didn't know that. I think using a slash is probably a bit clearer that it is part of this repo though.

Can you go through and remove those images though? It's a big problem for the git hash to be in the image name, not a tag (following a `:`). Good thing we removed the publishing :-)

It seems that in #26 we were publishing to a different place than the main build was, GHCR instead of ECR. It's good we deleted that, then, as there were two problems 1) publishing in soak tests and 2) publishing to two places instead of one. In hindsight, the soak tests PR was too huge to find these issues during review.

Anyways, preserving the code isn't a reason to do this - we publish where we want to publish, which so far has been ECR. Code is preserved in git history and it's not such a large amount either. Avoiding ECR pull limits (so being able to remove the docker login in the soak test flow, or running integration tests on PR builds, would be a reason to do so though so we could still go through with this change. But want to be on the same page first.

NathanielRN commented:
> Can you go through and remove those images though?

Yes you can. If you check [the link to the packages again](https://github.com/aws-observability/aws-otel-python/packages), I have deleted all the packages except for 1 for now. However, I would concede that it was rather cumbersome. It was like I was deleting a repository each time...

> In hindsight, the soak tests PR was too huge to find these issues during review.

I would agree that at 1,610 LOC #18 was rather large to review, however almost 70% of the code in that PR was because of the Python Scripts we added to replace the same thing we were doing with `bash` commands in the `soak-testing.yml` workflow file. I think it was a worthwhile change, but maybe the size is deceivingly large because these scripts which accomplish the same thing inflated the diff.

The plan to use GHCR in the Soak Tests was there since the beginning even when the PR was only 500 LOC. It was even something that came during the initial review but wasn't completely resolved: https://github.com/aws-observability/aws-otel-python/pull/18#discussion_r703147533. Perhaps we could have merged pieces 1 by 1 like the issue templates, then the docker-compose files, then workflow, etc... Either way like you said we fixed the solution and have a way forward without having done anything irreversible.

> Anyways, preserving the code isn't a reason to do this - we publish where we want to publish, which so far has been ECR.

Sorry I didn't mean to "merge it for the sake of preserving". I meant **I made this PR** for the sake of preserving code. We can let it sit in this PR state until we want it without a problem. I just didn't want to forget about how to make this change and I thought an open PR (or perhaps even a closed one) would be useful if we ever want something to refer to in the future.

> But want to be on the same page first.

Totally agreed. If you think this PR is too premature we can close it and refer back to it later whenever we want! 🙂

NathanielRN commented:
> It's a big problem for the git hash to be in the image name, not a tag (following a :).

I can definitely see your point here, I didn't realize container registries were so good at handling versions! I updated the PR to use the commit as version. The Soak Tests could work with this well so overall I definitely agree that this is the way to go.

anuraaga reviewed (approved):
I think switching to GHCR is fine due to the login requirement of ECR making it problematic for GitHub CI. Let's update the PR description to describe why we are switching to GHCR (which isn't really related to #26)

NathanielRN commented:
Updated the PR description to say this:

> Publishing to ECR is a problem because it requires an AWS account to publish the image. Further, viewing the images requires going to public.ecr.aws which is less obvious. In this PR we remove those pain points by using GHCR (GitHub Container Registry) which only requires a GITHUB_TOKEN with the packages: write permissions to upload images. Then, these images are accessible directly from the repo homepages for anyone interested in seeing more information about.
